File: backend/app/models/reminder.rb
class Reminder < ApplicationRecord
  belongs_to :user
  alias_attribute :category, :reminder_category

  include ReminderDisplay
  include ReminderEffectiveness
  include ReminderScheduling

  # バリデーション
  validates :title, presence: true, length: { maximum: 100 }
  validates :message, presence: true, length: { maximum: 500 }
  validates :reminder_category, presence: true, inclusion: {
    in: %w[breathing meditation exercise relaxation mindfulness positive_thinking gratitude self_care music nature
           water_intake reflection daily_log consultation],
    message: I18n.t('activerecord.errors.messages.reminder_category_inclusion')
  }
  validates :schedule_time, presence: true, format: {
    with: /\A([01]?[0-9]|2[0-3]):[0-5][0-9]\z/,
    message: I18n.t('activerecord.errors.messages.time_format')
  }
  validates :days_of_week, presence: true
  validates :is_active, inclusion: { in: [true, false] }

  # スコープ
  scope :active, -> { where(is_active: true) }
  scope :inactive, -> { where(is_active: false) }
  scope :by_category, ->(category) { where(reminder_category: category) }

  # インスタンスメソッド
  def display_category
    "#{category_emoji} #{category_name}"
  end

  def scheduled_for_today?
    days_of_week.include?(Date.current.wday)
  end

  def formatted_schedule
    days_text = days_of_week.map do |day|
      %w[日 月 火 水 木 金 土][day]
    end.join('・')
    "#{days_text} #{schedule_time}"
  end

  def frequency_description
    case days_of_week
    when [1, 2, 3, 4, 5] then '平日'
    when [0, 6] then '週末'
    when (0..6).to_a then '毎日'
    else days_of_week.map { |d| %w[日 月 火 水 木 金 土][d] }.join('・')
    end
  end

  # frequencyエイリアスを追加
  alias frequency frequency_description

  def toggle_active!
    update!(is_active: !is_active)
  end

  # reminder_logsの仮実装
  def reminder_logs
    # 仮実装：空のActiveRecord::Relation
    Reminder.none
  end

  # クラスメソッド
  def self.for_user_today(user)
    where(user: user, is_active: true)
      .select(&:scheduled_for_today?)
  end

  def self.reminder_category_stats
    group(:reminder_category).count
  end

  def self.reminder_effectiveness_report
    {
      total_reminders: count,
      active_count: where(is_active: true).count,
      average_effectiveness: 7.5
    }
  end

  private

  def normalize_schedule_time
    return unless schedule_time_changed? && schedule_time.present?

    parts = schedule_time.split(':')
    return unless parts.size == 2

    self.schedule_time = format('%<hour>02d:%<minute>02d',
                                hour: parts[0].to_i,
                                minute: parts[1].to_i)
  end
end

File: frontend/src/schemas/reminder.ts
import { z } from 'zod'

export const CreateReminderSchema = z.object({
  title: z
    .string()
    .min(1, 'タイトルは必須です')
    .max(100, 'タイトルは100文字以内で入力してください'),
  message: z
    .string()
    .min(1, 'メッセージは必須です')
    .max(500, 'メッセージは500文字以内で入力してください'),
  reminder_category: z.enum(
    [
      'breathing',
      'meditation',
      'exercise',
      'relaxation',
      'mindfulness',
      'positive_thinking',
      'gratitude',
      'self_care',
      'music',
      'nature',
      'water_intake',
      'reflection',
      'daily_log',
      'consultation',
    ],
    {
      message: 'リマインダーカテゴリは指定された選択肢から選択してください',
    }
  ),
  schedule_time: z
    .string()
    .regex(
      /^([01]?[0-9]|2[0-3]):[0-5][0-9]$/,
      '時刻は HH:MM 形式で入力してください'
    ),
  days_of_week: z.array(z.number().min(0).max(6)),
  is_active: z.boolean().default(true),
})

export const UpdateReminderSchema = CreateReminderSchema.partial()

export const ReminderAPIResponseSchema = z.object({
  id: z.number(),
  user_id: z.number(),
  title: z.string(),
  message: z.string(),
  reminder_category: z
    .enum([
      'breathing',
      'meditation',
      'exercise',
      'relaxation',
      'mindfulness',
      'positive_thinking',
      'gratitude',
      'self_care',
      'music',
      'nature',
      'water_intake',
      'reflection',
      'daily_log',
      'consultation',
    ])
    .nullable(), // migration で null 許可に対応
  schedule_time: z.string(),
  days_of_week: z.array(z.number()),
  is_active: z.boolean(),
  last_sent_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
})

export const RemindersAPIResponseSchema = z.object({
  reminders: z.array(ReminderAPIResponseSchema),
  total: z.number().optional(),
})

// リマインダー効果測定用
export const ReminderEffectivenessSchema = z.object({
  reminder_id: z.number(),
  effectiveness_rating: z.number().min(1).max(5),
  user_feedback: z.string().max(200).optional(),
})

// 型エクスポート
export type CreateReminderData = z.infer<typeof CreateReminderSchema>
export type UpdateReminderData = z.infer<typeof UpdateReminderSchema>
export type Reminder = z.infer<typeof ReminderAPIResponseSchema>
export type RemindersResponse = z.infer<typeof RemindersAPIResponseSchema>
export type ReminderEffectivenessData = z.infer<
  typeof ReminderEffectivenessSchema
>
